src/stdune/applicative_intf.ml

```diff
@@ -0,0 +1,9 @@
+module type S1 = sig
+  type 'a t
+
+  val pure : 'a -> 'a t
+
+  val ( let+ ) : 'a t -> ('a -> 'b) -> 'b t
+
+  val ( and+ ) : 'a t -> 'b t -> ('a * 'b) t
+end
```

Comment:
I suggest some changes to this interface:
* Rename `pure` to `return` (since our existing applicatives follow this convention).
* Add `map` and `map2` -- they are quite pervasive and will certainly be needed at some point.
* Move `let+` and `and+` into a sub-module `O` (again, our usual convention), and also to `Make` so that they can be obtained automatically.

Reply:
Isn't it a bit awkward to require all these helper functions? I.e. If I want to instantiate some code with a concrete applicative, I'd rather provide the smallest interface possible to it.

I don't have any strong preferences here, so feel free to push changes and tweak the interface as you wish

Reply:
Wouldn't it be a bit annoying to provide all these auxiliary functions when instantiating something like `String_with_vars.Make`? Or perhaps, I'm misunderstanding something. Feel free to push the changes you have in mind to this branch by the way. I don't feel too strongly about this stuff.

Reply:
I think eventually it's good to have (at least) two versions of applicative interface: a convenient one with at least `return`, `map`, `both`, `module O`, `all` etc., and a minimal one to pass to `Applicative.Make`. I think most applicative-generic code should take the convenient one, not the minimal one, so that the custom efficient versions  of things like `all` can be used.

In the end I don't think we'll need more than a few applicatives so the difficulty of implementing the "big" applicative interface doesn't frighten me.